### ISSUE TITLE
could not serialize access due to concurrent update

### ISSUE THREAD

**Issue by zredb** [frankhommers/Hangfire.PostgreSql#139]
the stack is:
 Npgsql.PostgresException (0x80004005): 40001: could not serialize access due to concurrent update
   at Npgsql.NpgsqlConnector.<>c__DisplayClass160_0.<<DoReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlConnector.<>c__DisplayClass160_0.<<DoReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlDataReader.<>c__DisplayClass41_0.<<ReadMessage>g__ReadMessageSequential|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming)
   at Npgsql.NpgsqlDataReader.NextResult()
   at Npgsql.NpgsqlCommand.ExecuteReaderAsync(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior)
   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)
   at Dapper.SqlMapper.ExecuteReaderWithFlagsFallback(IDbCommand cmd, Boolean wasClosed, CommandBehavior behavior) in C:\projects\dapper\Dapper\SqlMapper.cs:line 1053
   at Dapper.SqlMapper.QueryImpl[T](IDbConnection cnn, CommandDefinition command, Type effectiveType)+MoveNext() in C:\projects\dapper\Dapper\SqlMapper.cs:line 1081
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Dapper.SqlMapper.Query[T](IDbConnection cnn, String sql, Object param, IDbTransaction transaction, Boolean buffered, Nullable`1 commandTimeout, Nullable`1 commandType) in C:\projects\dapper\Dapper\SqlMapper.cs:line 723
   at Hangfire.PostgreSql.PostgreSqlJobQueue.<>c__DisplayClass5_1.<Dequeue_Transaction>b__1()
   at Hangfire.PostgreSql.Utils.Utils.TryExecute[T](Func`1 func, T& result, Func`2 smoothExValidator, Nullable`1 tryCount)
  Exception data:
    Severity: ERROR
    SqlState: 40001
    MessageText: could not serialize access due to concurrent update
    File: nodeModifyTable.c
    Line: 985
    Routine: ExecUpdate

**Comment by vytautask** [comment]
Somehow seems related to #73
Do you have a POC to illustrate this?

**Comment by Lenny415** [comment]
@SonicGD posted a POC on the original #73 issue:

https://github.com/frankhommers/Hangfire.PostgreSql/issues/73#issuecomment-582402084

**Comment by KrolikROJER** [comment]
Same problem.
may be **[skip locked](https://www.postgresql.org/docs/current/sql-select.html)** can help us

```
<PackageReference Include="Hangfire" Version="1.7.11" />
<PackageReference Include="Hangfire.PostgreSql" Version="1.6.4.2" />

```

```
2020-10-06 13:29:14.447 +04 [18016] ERROR:  could not serialize access due to concurrent update
2020-10-06 13:29:14.447 +04 [18016] STATEMENT:
        UPDATE "hangfire"."jobqueue"
        SET "fetchedat" = NOW() AT TIME ZONE 'UTC'
        WHERE "id" IN (
            SELECT "id"
            FROM "hangfire"."jobqueue"
            WHERE "queue" = ANY ($1)
            AND "fetchedat" IS NULL
            ORDER BY
            CASE "queue"
                WHEN 'motion' THEN 0
        WHEN 'exercises' THEN 1
        WHEN 'recordmotion' THEN 2
        WHEN 'outputparametersdto_v2' THEN 3
                ELSE 4
            END,
            "fetchedat", "jobid"
            LIMIT 1
        )
        RETURNING "id" AS "Id", "jobid" AS "JobId", "queue" AS "Queue", "fetchedat" AS "FetchedAt"
2020-10-06 13:29:44.551 +04 [20476] ERROR:  could not serialize access due to concurrent update
2020-10-06 13:29:44.551 +04 [20476] STATEMENT:
        UPDATE "hangfire"."jobqueue"
        SET "fetchedat" = NOW() AT TIME ZONE 'UTC'
        WHERE "id" IN (
            SELECT "id"
            FROM "hangfire"."jobqueue"
            WHERE "queue" = ANY ($1)
            AND "fetchedat" IS NULL
            ORDER BY
            CASE "queue"
                WHEN 'motion' THEN 0
        WHEN 'exercises' THEN 1
        WHEN 'recordmotion' THEN 2
        WHEN 'outputparametersdto_v2' THEN 3
                ELSE 4
            END,
            "fetchedat", "jobid"
            LIMIT 1
        )
        RETURNING "id" AS "Id", "jobid" AS "JobId", "queue" AS "Queue", "fetchedat" AS "FetchedAt"
2020-10-06 13:29:44.553 +04 [20476] ERROR:  could not serialize access due to concurrent update
2020-10-06 13:29:44.553 +04 [20476] STATEMENT:
        UPDATE "hangfire"."jobqueue"
        SET "fetchedat" = NOW() AT TIME ZONE 'UTC'
        WHERE "id" IN (
            SELECT "id"
            FROM "hangfire"."jobqueue"
            WHERE "queue" = ANY ($1)
            AND "fetchedat" IS NULL
            ORDER BY
            CASE "queue"
                WHEN 'motion' THEN 0
        WHEN 'exercises' THEN 1
        WHEN 'recordmotion' THEN 2
        WHEN 'outputparametersdto_v2' THEN 3
                ELSE 4
            END,
            "fetchedat", "jobid"
            LIMIT 1
        )
        RETURNING "id" AS "Id", "jobid" AS "JobId", "queue" AS "Queue", "fetchedat" AS "FetchedAt"
```

**Comment by shpikat** [comment]
Looks like it's the recommended approach, let me cite from the documentation:

> can be used to avoid lock contention with multiple consumers accessing a queue-like table

I believe performance-wise this approach is way better than classic `SERIALIZABLE` retries. Currently in our system we observe around 10-12 retries per single successful deque operation.

**Comment by frankhommers** [comment]
It's fixed in #169

**Comment by vytautask** [comment]
I've released it to nuget.org with version 1.8.0.
